utilities/titandb/titan_db_test.cc

```diff
@@ -361,6 +363,167 @@ TEST_F(TitanDBTest, DISABLED_ReadAfterDropCF) {
   }
 }
 
+TEST_F(TitanDBTest, BlobFileIOError) {
+  std::unique_ptr<FaultInjectionTestEnv> mock_env(
+      new FaultInjectionTestEnv(env_));
+  options_.env = mock_env.get();
+  Open();
+
+  std::map<std::string, std::string> data;
+  const int kNumEntries = 100;
+  for (uint64_t i = 1; i <= kNumEntries; i++) {
+    Put(i, &data);
+  }
+  ASSERT_EQ(kNumEntries, data.size());
+  CompactRangeOptions copts;
+  ASSERT_OK(db_->CompactRange(copts, nullptr, nullptr));
+  VerifyDB(data);
+
+  // not sync at reader
+  SyncPoint::GetInstance()->SetCallBack(
+    "BlobStorage::FindFile", [&](void *) {
+      mock_env->SetFilesystemActive(false, Status::IOError("Injected error"));
+    });
+  SyncPoint::GetInstance()->EnableProcessing();
+  for(auto& it : data) {
+    std::string value;
+    if(it.second.size() > options_.min_blob_size) {
+      ASSERT_FALSE(db_->Get(ReadOptions(), it.first, &value).ok());
+      mock_env->SetFilesystemActive(true);
+    }
+  }
+  std::unique_ptr<Iterator> iter(db_->NewIterator(ReadOptions()));
+  iter->SeekToFirst();
+  mock_env->SetFilesystemActive(true);
+  for (const auto& it : data) {
+    if(iter->Valid()) {
+      ASSERT_EQ(it.first, iter->key());
+      ASSERT_EQ(it.second, iter->value());
+      ASSERT_TRUE(it.second.size() < options_.min_blob_size);
+    } else {
+      ASSERT_TRUE(it.second.size() >= options_.min_blob_size);
+      break;
+    }
+    iter->Next();
+  }
+
+  options_.env = env_;
+  SyncPoint::GetInstance()->ClearAllCallBacks();
+  // env must be destructed AFTER db is closed to avoid
+  // `pure abstract method called` complaint.
+  iter.reset(nullptr); // early release to avoid outstanding reference
+  Close();
+  db_ = nullptr;
+}
+
+TEST_F(TitanDBTest, WriteIOErrorHandling) {
+  std::unique_ptr<FaultInjectionTestEnv> mock_env(
+      new FaultInjectionTestEnv(env_));
+  options_.env = mock_env.get();
+  Open();
+
+  std::map<std::string, std::string> data;
+  const int kNumEntries = 100;
+  for (uint64_t i = 1; i <= kNumEntries; i++) {
+    Put(i, &data);
+  }
+  ASSERT_EQ(kNumEntries, data.size());
+  CompactRangeOptions copts;
+  // no compaction to enable Flush
+  VerifyDB(data);
+
+  // test flush error handling
+  SyncPoint::GetInstance()->SetCallBack(
+    "FlushJob::Start", [&](void *) {
+      mock_env->SetFilesystemActive(false, Status::IOError("Injected error"));
+    });
+  SyncPoint::GetInstance()->EnableProcessing();
+  FlushOptions fopts;
+  ASSERT_TRUE(db_->Flush(fopts).IsIOError());
+  mock_env->SetFilesystemActive(true);
+  // subsequent writes return error too
+  WriteOptions wopts;
+  std::string key = "key_after_flush";
+  std::string value = "value_after_flush";
+  ASSERT_TRUE(db_->Put(wopts, key, value).IsIOError());
+  SyncPoint::GetInstance()->DisableProcessing();
+  SyncPoint::GetInstance()->ClearAllCallBacks();
+
+  // test compaction error handling
```

Comment:
The contract is, after write/flush/compaction getting IO error, the DB is not usable and need to be restart. Instead of restarting DB here, let's split compaction test to another test case.